### PR TITLE
BUG: Fix strange behaviour of Series.iloc on MultiIndex Series (#17148)

### LINKED ISSUE #17148: Series with MultiIndex gives surprising iloc results
Several puzzling behaviors appear below wherein calling `Series.iloc[int]` for a Series with a MultiIndex gives confusing results.  Namely, I see unexpected indices mutated.  Unexpected indices read and then added to others.  As well as more than one index being mutated.  Contrast all of the behaviors below with indexing a vanilla numpy array where none of the mystery occurs.  Full recreates below along with why I find them surprising.  I've seen similar behavior back in the Pandas 0.19.x series.

#### Code Sample 1

The length-2 case with duplicated keys at the first level.

```python
In [1]: import pandas as pd

In [2]: z = pd.DataFrame()
   ...: z['i'] = [1, 1]
   ...: z['j'] = [22, 33]
   ...: z['k'] = [555, 666]
   ...: z.set_index(['i', 'j'], inplace=True)
   ...: v = z.k
   ...: 

In [3]: v
Out[3]: 
i  j 
1  22    555
   33    666
Name: k, dtype: int64

In [4]: v.values
Out[4]: array([555, 666])

In [5]: v.iloc[0] += 200; v
Out[5]: 
i  j 
1  22    755
   33    666
Name: k, dtype: int64

In [6]: v.iloc[-1] += 300; v
Out[6]: 
i  j 
1  22    755
   33    966
Name: k, dtype: int64

In [7]: v.iloc[1] += 100; v
Out[7]: 
i  j 
1  22    1066
   33    1066
Name: k, dtype: int64
```

#### Code Sample 1 Problem / Expected Result

Final statement incremented key (1, 33) then wrote the result to be (1, 22) and (1, 33)?
Expected (1, 22) to remain 755 and (1, 33) to become 1066.

#### Code Sample 2

The length-2 case with unique keys at each level.

```python
In [1]: import pandas as pd

In [2]: z = pd.DataFrame()
   ...: z['i'] = [1, 2]
   ...: z['j'] = [3, 4]
   ...: z['k'] = [7, 8]
   ...: z.set_index(['i', 'j'], inplace=True)
   ...: v = z.k
   ...: 

In [3]: v
Out[3]: 
i  j
1  3    7
2  4    8
Name: k, dtype: int64

In [4]: v.values
Out[4]: array([7, 8])

In [5]: v.iloc[0] += 10; v
Out[5]: 
i  j
1  3    17
2  4     8
Name: k, dtype: int64

In [6]: v.iloc[-1] += 10; v
Out[6]: 
i  j
1  3    17
2  4    18
Name: k, dtype: int64

In [7]: v.iloc[1] += 1000; v
Out[7]: 
i  j
1  3    1018
2  4      18
Name: k, dtype: int64
```

#### Code Sample 2 Problem / Expected Result

Above iloc[0] incremented key (1, 3) but then iloc[1] does?
Do not expect 2 different iloc indices differing by 1 to point to the same location in a length 2 array.

#### Code Sample 3

The length-3 case.

```python
In [1]: import pandas as pd

In [2]: x = pd.DataFrame()
   ...: x['i'] = [1, 2, 3]
   ...: x['j'] = [11, 22, 33]
   ...: x['k'] = [4, 5, 6]
   ...: x.set_index(['i', 'j'], inplace=True)
   ...: v = x.k
   ...: 

In [3]: v
Out[3]: 
i  j 
1  11    4
2  22    5
3  33    6
Name: k, dtype: int64

In [4]: v.values
Out[4]: array([4, 5, 6])

In [5]: v.iloc[0] += 4; v
Out[5]: 
i  j 
1  11    8
2  22    5
3  33    6
Name: k, dtype: int64

In [6]: # Expected (3, 33) to be incremented but instead (2, 22) is?
   ...: v.iloc[2] += 7; v
   ...: 
Out[6]: 
i  j 
1  11     8
2  22    13
3  33     6
Name: k, dtype: int64

In [7]: # Expected (2, 22) to be incremented but instead (1, 11) is set to (2, 22) + increment?
   ...: v.iloc[1] += 10; v
   ...: 
Out[7]: 
i  j 
1  11    23
2  22    13
3  33     6
Name: k, dtype: int64
```

#### Code Sample 3 Problem / Expected Result

Comments inline above

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.1.35-pv-ts2
machine: x86_64
processor: 
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.utf8
LOCALE: en_US.UTF-8

pandas: 0.20.3
pytest: None
pip: 9.0.1
setuptools: 27.2.0
Cython: None
numpy: 1.13.1
scipy: None
xarray: None
IPython: 6.1.0
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by ante012**
- [x] closes #17148
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -329,6 +329,7 @@ Indexing
 - Fixes ``DataFrame.loc`` for setting with alignment and tz-aware ``DatetimeIndex`` (:issue:`16889`)
 - Avoids ``IndexError`` when passing an Index or Series to ``.iloc`` with older numpy (:issue:`17193`)
 - Allow unicode empty strings as placeholders in multilevel columns in Python 2 (:issue:`17099`)
+- Bug ``Series`` with ``MultiIndex`` gives surprising ``.iloc`` results when an integer indexer is used (:issue: `17148`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: needs to be a bit more explicit as this covers several cases.

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -146,7 +146,8 @@ def _get_setitem_indexer(self, key):
             return self._convert_tuple(key, is_setter=True)
 
         axis = self.obj._get_axis(0)
-        if isinstance(axis, MultiIndex):
+
+        if isinstance(axis, MultiIndex) and not isinstance(key, int):
```
Comment (jreback) [inline pandas/core/indexing.py]: use ``is_integer``

File: pandas/tests/indexing/test_iloc.py [hunk 3]
```diff
@@ -269,6 +269,14 @@ def test_iloc_setitem(self):
         expected = Series([0, 1, 0], index=[4, 5, 6])
         tm.assert_series_equal(s, expected)
 
+        df = pd.DataFrame(data=[[1, 2, 5], [1, 3, 6]], columns=['i', 'j', 'k'])
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: make a new test
add the issue number as a comment.

pls test all cases brought up.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 2 above)
Comment (jreback) [inline pandas/core/indexing.py]: also this should *only* be true if ``self.name == 'iloc'``, IOW we have a positional indexer.

File: pandas/tests/indexing/test_iloc.py (hunk 3 above)
Comment (ante328) [inline pandas/tests/indexing/test_iloc.py]: Thanks for the quick feedback! think its all fixed now

File: pandas/core/indexing.py (hunk 2 above)
Comment (ante328) [inline pandas/core/indexing.py]: Should I change `not is_integer(key)` to `self.name != 'iloc'`?

File: pandas/tests/indexing/test_iloc.py [hunk 4]
```diff
@@ -269,6 +269,40 @@ def test_iloc_setitem(self):
...
+
+            series = df.k.copy()
+            for i, v in zip(indexes, values):
+                series.iloc[i] += v
+
+            df.k = expected_k
+            expected = df.k.copy()
+            tm.assert_series_equal(series, expected)
+
+        check_scenario(
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: parametrize this test

File: pandas/core/indexing.py [hunk 5]
```diff
@@ -146,7 +146,8 @@ def _get_setitem_indexer(self, key):
             return self._convert_tuple(key, is_setter=True)
 
         axis = self.obj._get_axis(0)
-        if isinstance(axis, MultiIndex):
+
+        if isinstance(axis, MultiIndex) and not is_integer(key):
```
Comment (jreback) [inline pandas/core/indexing.py]: use ``self.name != 'iloc'``, and use a tests that fails this (IOW a MultiIndex with an integer first level that does not match the positions)

File: doc/source/whatsnew/v0.21.0.txt [hunk 6]
```diff
@@ -329,6 +329,7 @@ Indexing
 - Fixes ``DataFrame.loc`` for setting with alignment and tz-aware ``DatetimeIndex`` (:issue:`16889`)
 - Avoids ``IndexError`` when passing an Index or Series to ``.iloc`` with older numpy (:issue:`17193`)
 - Allow unicode empty strings as placeholders in multilevel columns in Python 2 (:issue:`17099`)
+- Bug in :func:`Series.iloc` when used with inplace addition or assignment and an int indexer on a ``MultiIndex`` ``Series`` causing the wrong indexes to be read from and written to (:issue:`17148`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: its a bug in ``.iloc`` not restricted to Series

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexing/test_iloc.py [hunk 7]
```diff
@@ -269,6 +269,30 @@ def test_iloc_setitem(self):
...
+            ([[1, 11, 4], [2, 22, 5], [3, 33, 6]], [0, -1, 1], [4, 7, 10],
+                [8, 15, 13])
+        ])
+    def test_iloc_setitem_int_multiindex_series(
+            self, data, indexes, values, expected_k):
+        # GH17148
+        df = pd.DataFrame(
+            data=data,
+            columns=['i', 'j', 'k'])
+        df.set_index(['i', 'j'], inplace=True)
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: ``df = df.set_index(...)``, don't use inplace in testing code.

File: pandas/tests/indexing/test_iloc.py [hunk 8]
```diff
@@ -269,6 +269,30 @@ def test_iloc_setitem(self):
...
+
+        series = df.k.copy()
+        for i, v in zip(indexes, values):
+            series.iloc[i] += v
+
+        df.k = expected_k
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: ``df['k'] = expected_k``

File: pandas/tests/indexing/test_iloc.py [hunk 9]
```diff
@@ -269,6 +269,30 @@ def test_iloc_setitem(self):
...
+        expected = df.k.copy()
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: this copy is necessary?

**Review by jreback — CHANGES_REQUESTED** [review]
does this hit all 3 cases in the issue? (can you enumerate them in the test)

**Comment by ante328** [comment]
I have been unable to recreate the error from the travis-ci test locally. The failed test doesn't hit the `_get_setitem_indexer` function when i run it here. Any tips on how to proceed are appreciated!

**Comment by ante328** [comment]
nvm is passes now

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**